dashboard/app/models/user.rb

```diff
@@ -2261,6 +2261,41 @@ def providers
     end
   end
 
+  # Returns number of times a user has attempted to join a section in the last 24 hours
+  # Returns nil if no section join attempts
+  def num_section_attempts
+    properties['section_attempts'] || 0
+  end
+
+  # There aare two possible states in which we would want to reset section attempts
+  # 1) Initialize for the first time 2) 24 hours have passed since last attempt
+  def reset_section_attempts?
+    !num_section_attempts || !properties['section_attempts_last_reset'] || num_section_attempts == 0 || (DateTime.now - DateTime.parse(properties['section_attempts_last_reset'])).to_i > 0
+  end
+
+  def reset_section_attempts
+    properties['section_attempts'] = 0
+    properties['section_attempts_last_reset'] = DateTime.now.to_s
+    save
+  end
+
+  def reset_section_attempts_if_needed
+    # Failed section attempts reset after 24 hours
+    reset_section_attempts if reset_section_attempts?
+  end
+
+  def display_captcha?
+    reset_section_attempts_if_needed
+    num_section_attempts >= 3
+  end
+
+  # Let's make sure we only hit this code when the user is already registered
+  def increment_section_attempts
+    reset_section_attempts_if_needed
+    properties.merge!({'section_attempts' => 1}) {|_, old_val, increment_val| old_val + increment_val}
+    save
```

Comment:
I'm wondering if here (and on 2279) we should save the object only if had previously been persisted.

One of the contributing factors in the original implementation is that `display_captcha?` eventually called `save(validate: false)` and that ended up being the first time the object was validated and persisted.  Because of the way the user model is written, some hooks only get run on the initial save.  It would be lower risk if this change did not change when that initial save occurs.